### PR TITLE
Add type definitions for expo 31 and expo 32

### PR DESCRIPTION
**Pull request by satya164**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/expo/expo/blob/master/CHANGELOG.md
- [x] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

If removing a declaration:
- [ ] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

**Comment by Freysi21** [comment]
This would really save my day @satya164

**Comment by PranavSenthilnathan** [comment]
typescript-bot thought this was a new definition. Adding the owners of the existing definition @KonstantinKai @martynaskadisa @janaagaard75 @ssanchezmarc@fhelwanger @umidbekkarimov @moshfeu @prokopcm @tinaroh @binki @mo @levansuper @ihmpavel @burtek @burtek @JKillian

**Comment by janaagaard75** [comment]
How do you use this with version 32 of Expo, @satya164? Since the `expo` packages now has it's own  type definitions, it's my experience that TypeScript selects those when writing `import ... from "expo"`. The closed pull request #31900 has some more info about this issue.

**Comment by satya164** [comment]
Ok, they have published a version without type declarations!

File: types/expo/index.d.ts [hunk 1]
```diff
@@ -2561,13 +2699,14 @@ export class Svg extends Component<{ width: number, height: number, viewBox?: st
  * Take Snapshot
  */
 export function takeSnapshotAsync(
-    view?: (number | React.ReactElement<any>),
+    node: number | React.ReactElement<any> | React.RefObject<any>,
```
Comment (janaagaard75) [inline types/expo/index.d.ts]: Why was this changed to `node`? [In the documentation it's still `view`](https://docs.expo.io/versions/latest/sdk/take-snapshot-async/#takesnapshotasyncview-options). Are the docs outdated?

File: types/expo/index.d.ts [hunk 2]
```diff
@@ -2561,13 +2699,14 @@ export class Svg extends Component<{ width: number, height: number, viewBox?: st
...
     options?: {
-        width?: number,
-        height?: number,
-        format?: 'png' | 'jpg' | 'jpeg' | 'webm',
-        quality?: number,
-        result?: 'file' | 'base64' | 'data-uri',
+        width?: number;
+        height?: number;
+        format: 'png' | 'jpg' | 'raw' | 'webm';
```
Comment (janaagaard75) [inline types/expo/index.d.ts]: `raw` is not listed in [the documentation](https://docs.expo.io/versions/latest/sdk/take-snapshot-async/#takesnapshotasyncview-options)?

Comment (janaagaard75) [inline types/expo/index.d.ts]: How about adding JSDoc comments for `format` noting that `png` is default and `webm` is only supported on Android?

File: types/expo/expo-tests.tsx [hunk 3]
```diff
@@ -33,24 +30,29 @@ import {
...
     Permissions,
     PublisherBanner,
-    Region,
     registerRootComponent,
     ScreenOrientation,
     SecureStore,
     Svg,
     Updates
 } from 'expo';
+import {
```
Comment (janaagaard75) [inline types/expo/expo-tests.tsx]: Sorry for asking all these questions, but why are these properties moved to `react-native-maps`? It's my understanding that `expo` contains everything that is needed, so I would not have include `react-native-maps` into my `package.json`.

**Comment by janaagaard75** [comment]
This pull request contains quite a lot of changes, @satya164. Would it be possible to break this into smaller pull requests, or at least more commits? That way it would be easier to get all the obvious changes committed right away.

**Comment by janaagaard75** [comment]
> Ok, they have published a version without type declarations!

Neat. What's the name of that package? Searching for `expo` on npmjs.com returns quite a lot of matches, and I couldn't find a note about this package in their [blog](blog.expo.io).

File: types/expo/index.d.ts (hunk 1 above)
Comment (satya164) [reply types/expo/index.d.ts]: @janaagaard75 I have copied most of the types here from their codebase. Some of the docs are outdated.

File: types/expo/index.d.ts (hunk 2 above)
Comment (satya164) [reply types/expo/index.d.ts]: https://github.com/expo/expo/pull/2771/files#diff-e1bb088e01a152f522c27cac3147cb34L9

Comment (satya164) [reply types/expo/index.d.ts]: I would have added them, but most of the types here don't have any comments and it's a lot of work to add comments for everything

File: types/expo/expo-tests.tsx (hunk 3 above)
Comment (satya164) [reply types/expo/expo-tests.tsx]: I moved them because there I had some error, but just moved them back. The expo definitions just re-export stuff from `react-native-maps`, so it's the same.

**Comment by satya164** [comment]
> Neat. What's the name of that package

It's the main expo package.

image

They just did it yesterday and it's not a new feature, so I doubt they will cover it in their blog post.

**Comment by satya164** [comment]
> Would it be possible to break this into smaller pull requests, or at least more commits?

How do you want me to break this? I can break it into 31 updates and 32, but I think it'll still be a lot to review because there will be a new file for 32 which is pretty big.

File: types/expo/index.d.ts (hunk 1 above)
Comment (janaagaard75) [reply types/expo/index.d.ts]: Thanks for the link.

File: types/expo/index.d.ts (hunk 2 above)
Comment (janaagaard75) [reply types/expo/index.d.ts]: Thanks for the link.

Comment (janaagaard75) [reply types/expo/index.d.ts]: Personally I think it would have been okay to just add a few documentation comments, but I certainly appreciate the willingly to include them everywhere. 🙂

File: types/expo/expo-tests.tsx (hunk 3 above)
Comment (janaagaard75) [reply types/expo/expo-tests.tsx]: 👍

**Review by janaagaard75 — APPROVED** [review]

**Comment by satya164** [comment]
Ok, I'll do it tomorrow.

**Comment by janaagaard75** [comment]
> Ok, I'll do it tomorrow.

No need to do it for this pull request. But it would be nice with smaller commits and smaller pull requests in the future.

**Comment by JKillian** [comment]
Thanks for doing this @satya164 / @janaagaard75! I upgraded to Expo 32 recently and noticed my types got noticeably worse in my app.  Using the path mapping to pick up these types fixes the issue though!
